### PR TITLE
OIDC active state form

### PR DESCRIPTION
**Pull request by th3coop**
This PR follows from #13980.  It adds form management portion of the ActiveState Publisher in PyPI Account settings.

Without #13980 merged, this PR have about 1400 more changed lines than it should.  It'll look more like this when 13980 is merged: https://github.com/ActiveState/warehouse/pull/3

### CODE REVIEW

**Comment by di** [comment]
I'll review this once #13980 is merged and conflicts are resolved here!

**Comment by th3coop** [comment]
@di, @woodruffw.  Before you give this PR a super thorough review, I'll point out that I've included quite a bit of duplication tests/unit/accounts/test_views.py.  I think I would be able to avoid this if I changed all of the validation tests in that file to call `_add_pending_oidc_publisher` and perhaps parameterize passing in the respective services values so I can check stuff like [this](https://github.com/ActiveState/warehouse/blob/main/warehouse/accounts/views.py#L1533) has the right service name in the string?

That could result in quite large or complex parameterize lists though.

Any thoughts?

**Comment by di** [comment]
@th3coop We already have parametrized tests for these which were added recently, e.g. for that specific example: https://github.com/pypi/warehouse/blob/e62049b47483464239fc35df01854274da16c091/tests/unit/accounts/test_views.py#L3442-L3526

I think if you get this branch up to date, you should just be able to add a new ActiveState parameter to these existing tests, rather than duplicating them.

**Comment by th3coop** [comment]
@di, i was expecting these merge conflicts to go away when https://github.com/pypi/warehouse/pull/13980 was merged since I did a fresh rebase yesterday but I still see them now that that code is in `main`.  I'll rebase again.  I am now realizing that those are the params you already told me about a couple days ago.

**Comment by di** [comment]
@th3coop Is it possible you're rebasing on the out-of-date `main` from the Activestate fork rather than from PyPI's remote?

**Comment by th3coop** [comment]
@di, that is possible.  I often pull in the upstream `main` to my local environment though.  And I was certain I had done that before doing my last series of rebases but I could easily forgotten.  Sorry for the delay on this one.  I really thought it was ready for review already.  I'm fixing these merge conflicts right now.

**Comment by th3coop** [comment]
OK.  @di,I got things fixed up.  Again, sorry about that mixup.  Also, I'm working on getting **Allowing changes to a pull request branch created from a fork** enabled on this PR.  I don't appear to have permissions to do that.

**Comment by th3coop** [comment]
Hey @di.  It appears that we can't turn on **Allowing changes to a pull request branch created from a fork
**.  The option to enable that doesn't show up for myself or any admins of our GitHub account and I can't find anything on the internet as to why that might be.  I'm like 75% sure that it's because our fork is in the ActiveState org and not my personal GitHub account.  That's my best guess based on step 4 [from the docs](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests) where it says **On user-owned forks**.

If not having this turned on is a huge pain for y'all, let me know and I can move the remaining PRs to my personal account.

**Comment by woodruffw** [comment]
Yeah, that's an unfortunate known long-standing bug with GitHub orgs and forks. The hacky workaround is to manually give the person commit rights on your fork, but that's not always desirable :upside_down_face:

**Comment by th3coop** [comment]
I'm fine with that! Thanks for the workaround @woodruffw.  Invites sent to yourself and DI.  I'll remove you both when we get the PRs merged.

File: warehouse/oidc/forms/activestate.py
Comment (di) [inline warehouse/oidc/forms/activestate.py:26]: Just sanity checking these... `---` or `-foo-` is a valid "org url" name? `...` or `-.-` is a valid project name?

File: warehouse/oidc/forms/activestate.py
Comment (di) [inline warehouse/oidc/forms/activestate.py:73]: Should the org regex be reused here?

File: warehouse/oidc/forms/activestate.py
Comment (di) [inline warehouse/oidc/forms/activestate.py:118]: Seems like we could probably condense these into:

```python
except (requests.HTTPError, requests.ConnectionError, request.Timeout) as e:
    sentry_sdk.capture_message(
        f"Unexpected error {e!r} from ActiveState organization lookup: {response.content!r}"
    )
```

File: warehouse/oidc/forms/activestate.py
Comment (di) [inline warehouse/oidc/forms/activestate.py:146]: Any reason not to just put the body of this function here instead?

File: warehouse/oidc/forms/activestate.py
Comment (di) [inline warehouse/oidc/forms/activestate.py:214]: Ditto here.

File: warehouse/oidc/forms/activestate.py
Comment (di) [inline warehouse/oidc/forms/activestate.py:241]: This should use the `PendingPublisherMixin` instead.

File: warehouse/oidc/forms/activestate.py
Comment (th3coop) [reply warehouse/oidc/forms/activestate.py:26]: They are not.  I've missed a few validations here.  Fixing and adding a couple more test cases.

File: warehouse/oidc/forms/activestate.py
Comment (th3coop) [reply warehouse/oidc/forms/activestate.py:73]: Yes, we use the same validation on usernames.  Maybe I should make that explicit and name this in a way that it's clear that it's meant to be for both.  Would you prefer that?

File: warehouse/oidc/forms/activestate.py
Comment (th3coop) [reply warehouse/oidc/forms/activestate.py:118]: This is so nice.  Thanks!

File: warehouse/oidc/forms/activestate.py
Comment (th3coop) [reply warehouse/oidc/forms/activestate.py:146]: No dice on this.  Some step in the code path of looking up validation functions based on the field name causes stubbing not to work.  When I try to stub `validate_organization`, the original code runs instead of the stubbed function.  If I keep `_lookup_organization` the stub works.

File: warehouse/oidc/forms/activestate.py
Comment (woodruffw) [inline warehouse/oidc/forms/activestate.py:87]: Nitpick: unnecessary f-string:

```suggestion
                ACTIVESTATE_GRAPHQL_API_URL,
```

File: warehouse/oidc/forms/activestate.py
Comment (woodruffw) [inline warehouse/oidc/forms/activestate.py:153]:
```suggestion
                ACTIVESTATE_GRAPHQL_API_URL,
```

File: warehouse/oidc/forms/activestate.py
Comment (woodruffw) [inline warehouse/oidc/forms/activestate.py:87]: Nitpick, nonblocking: given that this `POST` is made both here and in `_lookup_actor` and has mostly identical error handling in both cases, maybe we can abstract it into a helper function? Something like:

```python
def _activestate_request(query: dict, variables: dict) -> dict: ...
```

(OTOH, this might make the error states harder to handle, so maybe not worth it.)

File: warehouse/oidc/forms/activestate.py
Comment (woodruffw) [inline warehouse/oidc/forms/activestate.py:33]: Nitpick, nonblocking: consider making these `_ACTIVESTATE_GRAPHQL_API_URL`, etc. since they only get used in this module :slightly_smiling_face:

**Review by woodruffw — COMMENTED** [review]
A couple of small nitpicks; otherwise this is looking really good to me! I'm going to stand it up locally tomorrow and do some test-driving as well.

File: warehouse/oidc/forms/activestate.py
Comment (th3coop) [inline warehouse/oidc/forms/activestate.py:94]: I've removed this and instead am doing a check like this:
`if response.status_code >= 400:`
because I'm "UnboundLocalError" when a TimeOut error occurs. This arose when i collapsed all the HTTPError handling into 1 `except` line.

File: warehouse/oidc/forms/activestate.py
Comment (th3coop) [reply warehouse/oidc/forms/activestate.py:73]: I made it clearer what this can be used for.

File: warehouse/oidc/forms/activestate.py
Comment (th3coop) [reply warehouse/oidc/forms/activestate.py:87]: Done locally.

Comment (th3coop) [reply warehouse/oidc/forms/activestate.py:87]: This worked out nicely.  I'm using a "response handler" function to do handle the unique form code paths where a clear validation error needs to be displayed.

**Comment by th3coop** [comment]
@woodruffw.  All set.

File: warehouse/oidc/forms/activestate.py
Comment (woodruffw) [inline warehouse/oidc/forms/activestate.py:28]: Typo?

```suggestion
_DOUBLE_DASHES = re.compile(r"--+")
```

Comment (th3coop) [reply warehouse/oidc/forms/activestate.py]: Doh...Indeed. Fixed locally to update where it's used as well.

Comment (th3coop) [reply warehouse/oidc/forms/activestate.py]: Pushed.

File: tests/unit/oidc/forms/test_activestate.py
Comment (th3coop) [inline tests/unit/oidc/forms/test_activestate.py:153]: Patched this lower case "API" to be upper case.

File: tests/unit/oidc/forms/test_activestate.py
Comment (th3coop) [inline tests/unit/oidc/forms/test_activestate.py:383]: Remove unnecesssary formatting string and uppercased API

**Review by woodruffw — APPROVED** [review]
LGTM overall! I didn't get to the local deployment today, but will test tomorrow.

File: warehouse/accounts/views.py
Comment (woodruffw) [inline warehouse/accounts/views.py:1533]: No action required, but this is making me realize that we should probably get rid of these hardcoded provider names and replace them with something in `oidc.utils`.

File: warehouse/manage/views/__init__.py
Comment (woodruffw) [inline warehouse/manage/views/__init__.py:1537]: I think `publisher_url()` can never be falsey here, right? Based on:

https://github.com/pypi/warehouse/blob/30c8720951db492dbd800bb5433edbdf1fc6085f/warehouse/oidc/models/activestate.py#L125-L128

So, I think this can be:

```suggestion
            f"Added {publisher} "
            + f"in {publisher.publisher_url()} "
            + f"to {self.project.name}",
```

(and then maybe reduce the lines?)

File: warehouse/oidc/forms/activestate.py
Comment (woodruffw) [inline warehouse/oidc/forms/activestate.py:35]: Nitpick, nonblocking: I think `_TIMEOUT` is now only used in one place, so we can just drop the constant.

File: warehouse/oidc/forms/activestate.py
Comment (woodruffw) [inline warehouse/oidc/forms/activestate.py:58]: Thought, nonblocking: I think this could be worked into the regexps themselves, rather than having a separate validator. But IMO this is not worth trying to do now, something we can iterate on in the future 🙂

**Review by woodruffw — COMMENTED** [review]
One change request and a few nitpicks that I leave up to you!

This is looking really good!

File: warehouse/manage/views/__init__.py
Comment (th3coop) [reply warehouse/manage/views/__init__.py:1537]: Oh man, good eye.

File: warehouse/accounts/views.py
Comment (th3coop) [reply warehouse/accounts/views.py:1533]: I'm going to leave this one out as there's a bit (very little) of nuance to what the change could be exactly.

File: warehouse/oidc/forms/activestate.py
Comment (th3coop) [reply warehouse/oidc/forms/activestate.py:58]: I think I'm going to skip this as I actually prefer if we had more isolated validators that can give specific feedback.  The current regex, if it were to give helpful feedback, needs to state multiple things as the possible fix.  With this specific validator, it can state exactly what's wrong.

File: warehouse/manage/views/__init__.py
Comment (th3coop) [reply warehouse/manage/views/__init__.py:1537]: Having a bit of trouble with this change.  It looks like the [Google Publisher Model](https://github.com/ActiveState/warehouse/blob/main/warehouse/oidc/models/google.py#L84) returns None for `publisher_url` and I'm not, if that can even be changed, what it would be.

Mind if I leave this change out as well?

Comment (th3coop) [reply warehouse/manage/views/__init__.py:1537]: Actually, this points out an issue in my code.  In my attempted to be type safe, i'm breaking the interface we're using for the PublisherMixin.  I'm going to change `publisher_url` in the AtiveState Publisher mixin to not have the type declarations I added.  This can be iterated on later and done across all published to make sure they match.

Comment (th3coop) [reply warehouse/manage/views/__init__.py]: OK, sorry for the noise.  I DID make this change but I also backed out my additional type declarations that make the `ActiveStatePublisherMixin.publisher_url` not match what the others `*PublisherMixin.publisher_url` have.

**Comment by th3coop** [comment]
Okie dokie.  All changes pushed, @woodruffw.

**Comment by woodruffw** [comment]
Thanks! I'm taking another look today.

(You'll need to update against `main` for the merge as well.)

File: warehouse/oidc/forms/activestate.py
Comment (woodruffw) [inline warehouse/oidc/forms/activestate.py:158]: I think this can be removed outright, since we're not specializing the ctor in any way:

```suggestion
```

File: warehouse/manage/views/__init__.py
Comment (th3coop) [reply warehouse/manage/views/__init__.py:1541]: There's been a couple of these "great but not right now" comments.  How do you all track things like this?  Should I file Enhancement request Issue reports in the Warehouse repo or do you have some other method of tracking possible future work?

File: warehouse/oidc/forms/activestate.py
Comment (th3coop) [reply warehouse/oidc/forms/activestate.py:76]: This `noqa` can probably go away.  I'm fairly certain I had added it due to a much longer msg string.  I'll remove.

Comment (th3coop) [reply warehouse/oidc/forms/activestate.py]: Done.  Also reformatted the line to split the line up but keep it as one line when output.

**Comment by th3coop** [comment]
Thanks so much for all your time @woodruffw !

**Review by woodruffw — APPROVED** [review]

**Comment by th3coop** [comment]
@miketheman will you have time to review this PR this week?

File: warehouse/oidc/forms/activestate.py
Comment (miketheman) [inline warehouse/oidc/forms/activestate.py:193]: I was reading more about wtforms validators, and I _think_ that if there's a named validator function, it fires **before** the validators on the field, so `validate_actor()` would perform API calls before the other validators like regex/dashes are in play.

It's definitely worth re-verifying - since if our validators operate in that order, then we're performing an API call regardless. I hope I'm wrong! If that _is_ true, then it's not only this form - so I don't think it's immediately actionable, but if you can test it, that'd be great.

File: warehouse/accounts/views.py
Comment (miketheman) [inline warehouse/accounts/views.py:1523]: **suggestion:** use the code we're disabling so that we have a better idea ofwhy a rule was disabled.‏

```suggestion
            "pending_activestate_publisher_form": self.pending_activestate_publisher_form,  # noqa: E501
```

Similar note applies to other `noqa` instances.

**Review by miketheman — CHANGES_REQUESTED** [review]
Thanks to the heavy lifting the other folks did before, this was relatively easy to review.

I had a few notes, questions, comments inline - the only real blocker are with the linter disables which should be an easy fix.

There's other questions and comments inline, feel free to address them as you wish.
The branch should also be rebased/merged once ready.

**Comment by th3coop** [comment]
Thanks for the review, @miketheman!  Great notes in there.  I'll work through them either later today or tomorrow.  The GQL API is available again if you wanted to, or have time to take it for a manual test drive.  Again, sorry for that terrible timing.

File: warehouse/templates/manage/account/publishing.html
Comment (th3coop) [reply warehouse/templates/manage/account/publishing.html:250]: Would it be more clear if I changed the wording slightly to something like:
 
> The username of the user that triggered the build in the ActiveState Platform that will publish your package
 
There is certainly contextual wording in the help text here but it's context I would expect someone that is configuring this to have.  eg. someone that uses the ActiveState Platform to manage their Python installs knows what it means for a user to trigger a build of their project, or even what a project is.

The ActiveState docs that explain the steps to build your package will elucidate this as well.

To actually answer your question: The only links I can think of that would clarify this help string would be a link to our "Getting Started" docs here: https://docs.activestate.com/platform/start/getting-started/

I think the help link you'd be interested would be much more focused than that.

File: warehouse/oidc/models/activestate.py
Comment (th3coop) [reply warehouse/oidc/models/activestate.py:125]: Curious about why I removed the type declarations?  Or why claims are passed to this function at all?  

The former I did because, at the time, `oidc.models._core.OIDCPublisherMixin.publisher_url` didn't have the types.  Now that I look again though, they DO have them, though slightly different.  I'll bring them back.

The latter:  To fulfill the "interface" of `OIDCPublisherMixin` and override `pubslisher_url` there.  I could technically leave it out but we could run into issues some day if someone is iterating over publishers and calling their `publisher_url` and it fails on the ActiveState publisher because `publisher_url` only takes `self` and the calling code is trying to pass it `claims`.

I'll push a changes to bring back the typing.  Let me know if my explanation of including `claims` as a method arg makes sense.

File: warehouse/oidc/forms/activestate.py
Comment (th3coop) [reply warehouse/oidc/forms/activestate.py:34]: Looks like I was using `Try again in a few minutes`, like the GitHub form, before.  I'll bring that back in the contexts that make sense.

File: warehouse/oidc/forms/activestate.py
Comment (th3coop) [reply warehouse/oidc/forms/activestate.py:57]: I don't think I totally agree here.  I'm providing the granular regex's checks to provide a better user experience in this particular context but we could get away with "That organization does not exist".  Then the user would double check what they entered.

Maybe that's the better solution.  The user isn't trying to come up with an org name, they know the org.  They just need feedback that they entered the wrong thing, ie. "Not found"

So I would remove the regex checks on the format of the org name, then it falls to `_lookup_organization` then the empty list response is caught there.

What do you think @miketheman or @woodruffw ?

File: warehouse/oidc/forms/activestate.py
Comment (th3coop) [reply warehouse/oidc/forms/activestate.py:193]: If you like the idea here https://github.com/pypi/warehouse/pull/15168#discussion_r1480490211. I would do that change for this validator as well.  We really just need a "you entered the wrong text" kind of feedback here.

File: warehouse/accounts/views.py
Comment (th3coop) [reply warehouse/accounts/views.py:1523]: Done this in a local change.  Nice catch.  Didn't notice that used elsewhere in the code.  I like this much better!

File: warehouse/templates/manage/account/publishing.html
Comment (miketheman) [inline warehouse/templates/manage/account/publishing.html:203]: I hate to do this, but I just merged #10003 that uses newlines to make these longer lines more readable. Would you mind reflowing these in the rest of the changed template code?

```suggestion
      {{ pending_activestate_pubisher_form.project_name(
          placeholder=gettext("project name"),
          autocomplete="off",
          autocapitalize="off",
          spellcheck="false",
          class_="form-group__field",
          aria_describedby="project_name-errors"
      ) }}
```

File: warehouse/templates/manage/account/publishing.html
Comment (miketheman) [inline warehouse/templates/manage/account/publishing.html:233]: nit: move `aria-describedby` out of a splat at to a proper keyword `aria_describedby`

Similar note applies to the rest of the templates.

**Review by miketheman — CHANGES_REQUESTED** [review]
@th3coop we're almost there! I dropped a couple of notes inline, should be straightforward changes.

Then a rebase and `make translations` update, and should be good to merge!

File: warehouse/templates/manage/account/publishing.html
Comment (th3coop) [reply warehouse/templates/manage/account/publishing.html:203]: MIIIiiiIIiiIIIiIIiiIIiiIIIIIke (pouty face emoji and said with a whiny lilt)

Comment (th3coop) [reply warehouse/templates/manage/account/publishing.html:203]: Kidding.  I prefer the new format muchly.  I'll do that for sure.

File: warehouse/manage/views/__init__.py
Comment (th3coop) [reply warehouse/manage/views/__init__.py:1541]: @miketheman, I'm not sure if you're saying we should do that in this PR or should do that after this PR.  If it's the former, I'd prefer not to mix those changes into this PR.  If it's the latter, I'd be interested in doing that for your team with a follow up PR.

Comment (miketheman) [reply warehouse/manage/views/__init__.py:1541]: Sorry for any confusion! Leave it alone for now.

File: warehouse/templates/manage/account/publishing.html
Comment (th3coop) [reply warehouse/templates/manage/account/publishing.html:233]: oh wow....good eye!

**Comment by miketheman** [comment]
> I'm leaning towards removing them though.

@th3coop go ahead and remove them - less is more 😁

**Comment by th3coop** [comment]
> go ahead and remove them - less is more 😁

Agreed and done.  Fixed the tests and took it for a spin and it works as I'd expect it to

**Review by miketheman — CHANGES_REQUESTED** [review]
Almost there! just a couple last bits, and rebase against `main`, and I think we're good to do.

File: warehouse/oidc/forms/activestate.py
Comment (th3coop) [reply warehouse/oidc/forms/activestate.py]: Huh, those are both moderately embarrassing.  Sorry about those misses, Mike.

**Review by miketheman — APPROVED** [review]
Yay! I'll approve now, and merge in the AM.

**Comment by th3coop** [comment]
> Yay! I'll approve now, and merge in the AM.

Hurray!  Thanks so much Mike, Will and Dustin!

**Comment by miketheman** [comment]
@th3coop I've gone ahead and enabled the OIDC provider on TestPYI so you can test run it over there.

**Pull request merged**
